File: python/tests/test_wire_recordings.py
"""wire/recordings against a fake device: a command runner plus a real loopback
TCP sender that behaves like the firmware's (connect, bytes, close)."""
from __future__ import annotations

import socket
import threading

import pytest

from visio_schema.v1.control import command_result_pb2
from visio_schema.wire import recordings as rec

SESSION = "session_00042-1789017895"


class FakeDevice:
    """Serves ListRecordings / OpenRecordingFile / DeleteRecording, and sends
    opened ranges on a loopback socket exactly as the firmware does: one
    connection per open, bytes, close. ``cuts`` makes the next sends stop
    short; ``refuse_sends`` closes every connection without sending."""

    def __init__(self, files: dict[str, bytes], *, sessions: list[str] | None = None,
                 page: int = 2):
        self.files = dict(files)
        self.mtime = {name: 1_789_017_895_000_000_000 + i for i, name in enumerate(files)}
        self.sessions = sessions or [SESSION]
        self.page = page
        self.opens: list = []
        self.cuts: list[int] = []
        self.refuse_sends = False
        self.errors: dict[str, str] = {}
        self._pending = None
        self._server = socket.create_server(("127.0.0.1", 0))
        self.port = self._server.getsockname()[1]
        self._thread = threading.Thread(target=self._serve, daemon=True)
        self._thread.start()

    def close(self):
        self._server.close()

    def _serve(self):
        while True:
            try:
                conn, _ = self._server.accept()
            except OSError:
                return
            with conn:
                pending, self._pending = self._pending, None
                if pending is None or self.refuse_sends:
                    continue
                data, offset, length = pending
                chunk = data[offset:offset + length]
                if self.cuts:
                    chunk = chunk[:self.cuts.pop(0)]
                try:
                    conn.sendall(chunk)
                    conn.shutdown(socket.SHUT_WR)
                except OSError:  # the host hung up mid-send, as a failing sink does
                    pass

    def run(self, cmd, timeout):
        result = command_result_pb2.CommandResult(command_id=cmd.command_id, ok=True)
        body = cmd.WhichOneof("body")
        if body in self.errors:
            result.ok = False
            result.error_code = self.errors[body]
            return result
        if body == "list_recordings":
            req = cmd.list_recordings
            out = result.recordings
            if req.session_name:
                entry = out.recordings.add(name=req.session_name, file_count=len(self.files))
                for name, data in self.files.items():
                    entry.files.add(name=name, size=len(data), mtime_ns=self.mtime[name],
                                    complete=True)
                return result
            start = int(req.cursor) if req.cursor else 0
            for name in self.sessions[start:start + self.page]:
                out.recordings.add(name=name)
            if start + self.page < len(self.sessions):
                out.next_cursor = str(start + self.page)
            return result
        if body == "open_recording_file":
            req = cmd.open_recording_file
            self.opens.append((req.offset, req.expect_size, req.expect_mtime_ns))
            data = self.files[req.file_name]
            mtime = self.mtime[req.file_name]
            if (req.expect_size or req.expect_mtime_ns) and \
                    (req.expect_size, req.expect_mtime_ns) != (len(data), mtime):
                result.ok = False
                result.error_code = "changed"
                return result
            length = len(data) - req.offset
            self._pending = (data, req.offset, length)
            result.file_open.CopyFrom(command_result_pb2.RecordingFileOpen(
                port=self.port, offset=req.offset, length=length,
                file_size=len(data), mtime_ns=mtime))
            return result
        if body == "delete_recording":
            self.sessions.remove(cmd.delete_recording.session_name)
            return result
        result.ok = False
        result.error_code = "unsupported"
        return result


@pytest.fixture
def device():
    dev = FakeDevice({"ego_0000.mcap": bytes(range(256)) * 4096, "session.json": b"{}"})
    yield dev
    dev.close()


class Collector:
    def __init__(self):
        self.data = bytearray()
        self.offsets: list[int] = []

    def __call__(self, offset: int, chunk: bytes):
        assert offset == len(self.data), "sink offsets must be contiguous"
        self.offsets.append(offset)
        self.data += chunk


# ---- builders -----------------------------------------------------------------

@pytest.mark.parametrize("bad", ["", ".hidden", "a/b", "..", "x" * 64, "nul\0"])
def test_open_refuses_names_the_device_cannot_accept(bad):
    with pytest.raises(ValueError):
        rec.open_recording_file_command(bad, "ego_0000.mcap")
    with pytest.raises(ValueError):
        rec.open_recording_file_command(SESSION, bad)


def test_a_name_at_the_cap_is_accepted():
    rec.open_recording_file_command("s" * rec.MAX_NAME_BYTES, "f" * rec.MAX_NAME_BYTES)


def test_delete_requires_a_target():
    with pytest.raises(ValueError):
        rec.delete_recording_command(SESSION, target_device="")


def test_plain_list_is_the_original_call():
    cmd = rec.list_recordings_command()
    assert cmd.WhichOneof("body") == "list_recordings"
    # field 18, length-delimited, empty: exactly what pre-pull hosts send.
    assert cmd.SerializeToString() == bytes.fromhex("920100")


# ---- listing ------------------------------------------------------------------

def test_list_all_follows_cursors_to_the_last_page():
    dev = FakeDevice({}, sessions=[f"session_{i:05d}-1" for i in range(5)], page=2)
    try:
        names = [e.name for e in rec.list_all_recordings(dev.run)]
    finally:
        dev.close()
    assert names == [f"session_{i:05d}-1" for i in range(5)]


def test_a_cursor_that_does_not_advance_is_a_protocol_error():
    def run(cmd, timeout):
        result = command_result_pb2.CommandResult(ok=True)
        result.recordings.next_cursor = "same"
        return result
    with pytest.raises(rec.RecordingsError) as e:
        rec.list_all_recordings(run)
    assert e.value.code == "protocol"


def test_list_session_files(device):
    entry = rec.list_session_files(device.run, SESSION)
    assert [f.name for f in entry.files] == ["ego_0000.mcap", "session.json"]


# ---- pulling ------------------------------------------------------------------

def test_pull_whole_file_in_one_open(device):
    sink = Collector()
    opened = rec.pull_file(device.run, "127.0.0.1", SESSION, "ego_0000.mcap", sink)
    assert bytes(sink.data) == device.files["ego_0000.mcap"]
    assert opened.file_size == len(device.files["ego_0000.mcap"])
    assert len(device.opens) == 1


def test_an_early_close_resumes_at_the_bytes_received_with_identity_pinned(device):
    device.cuts = [100_000, 300_000]
    sink = Collector()
    rec.pull_file(device.run, "127.0.0.1", SESSION, "ego_0000.mcap", sink)
    assert bytes(sink.data) == device.files["ego_0000.mcap"]
    size, mtime = len(device.files["ego_0000.mcap"]), device.mtime["ego_0000.mcap"]
    assert device.opens == [(0, 0, 0), (100_000, size, mtime), (400_000, size, mtime)]


def test_a_file_that_changes_between_opens_fails_instead_of_splicing(device):
    device.cuts = [50_000]
    sink = Collector()
    real_run = device.run

    def run(cmd, timeout):
        if cmd.WhichOneof("body") == "open_recording_file" and device.opens:
            device.mtime["ego_0000.mcap"] += 1
        return real_run(cmd, timeout)

    with pytest.raises(rec.RecordingsError) as e:
        rec.pull_file(run, "127.0.0.1", SESSION, "ego_0000.mcap", sink)
    assert e.value.code == "changed"
    assert len(sink.data) == 50_000


def test_opens_that_deliver_nothing_give_up(device):
    device.refuse_sends = True
    with pytest.raises(rec.RecordingsError) as e:
        rec.pull_file(device.run, "127.0.0.1", SESSION, "ego_0000.mcap", Collector(),
                      max_opens_without_progress=3)
    assert e.value.code == "no_progress"
    assert len(device.opens) == 3


def test_an_already_complete_file_needs_no_socket(device):
    size = len(device.files["session.json"])
    opened = rec.pull_file(device.run, "127.0.0.1", SESSION, "session.json", Collector(),
                           offset=size)
    assert opened.length == 0


def test_device_refusals_carry_the_device_code(device):
    device.errors["open_recording_file"] = "writing"
    with pytest.raises(rec.RecordingsError) as e:
        rec.pull_file(device.run, "127.0.0.1", SESSION, "ego_0000.mcap", Collector())
    assert e.value.code == "writing"


def test_an_open_answer_for_another_offset_is_a_protocol_error(device):
    def run(cmd, timeout):
        result = device.run(cmd, timeout)
        result.file_open.offset += 1
        return result
    with pytest.raises(rec.RecordingsError) as e:
        rec.open_recording_file(run, SESSION, "ego_0000.mcap")
    assert e.value.code == "protocol"


def test_a_stalled_socket_returns_what_arrived_with_the_reason():
    server = socket.create_server(("127.0.0.1", 0))
    held = []
    threading.Thread(target=lambda: held.append(server.accept()), daemon=True).start()
    opened = command_result_pb2.RecordingFileOpen(port=server.getsockname()[1], length=10,
                                                 file_size=10)
    got = rec.receive("127.0.0.1", opened, Collector(), stall_timeout=0.2)
    server.close()
    assert got.received == 0
    assert isinstance(got.error, TimeoutError)


def test_sink_failures_propagate(device):
    def sink(offset, chunk):
        raise OSError(28, "No space left on device")
    with pytest.raises(OSError):
        rec.pull_file(device.run, "127.0.0.1", SESSION, "ego_0000.mcap", sink)


def test_delete(device):
    rec.delete_recording(device.run, SESSION, target_device="GILABS-A")
    assert device.sessions == []


def test_pull_quiesce_rules_are_the_spec_section_7_list():
    # The C++ (kPullQuiesceRules) and TS (PULL_QUIESCE_RULES) twins pin the same list.
    assert rec.PULL_QUIESCE_RULES == ("**/camera/*", "**/imu/*/raw", "**/imu/*/quat", "**/audio/*")

File: cpp/tests/test_wire_recordings.cc
/*
 * test_wire_recordings.cc - the C++ recordings-pull client.
 *
 * Two halves: replay tests/golden/recordings_wire_vectors.txt (the same file the
 * Python and TypeScript suites replay, so a codec drift shows up as a byte diff),
 * and drive PullFile against a loopback sender that behaves like the firmware's
 * (one connection per open, bytes, close).
 */
#include "visio_schema/wire/recordings.hpp"

#include <arpa/inet.h>
#include <netinet/in.h>
#include <sys/socket.h>
#include <unistd.h>

#include <atomic>
#include <cstring>
#include <map>
#include <mutex>
#include <string>
#include <thread>
#include <vector>

#include <gtest/gtest.h>

#include "golden_vectors_test_util.hpp"

namespace rec = visio_schema::wire::recordings;

namespace {

const char* kSession = "session_00042-1789017895";
const char* kCursor = "v1:1789017895000000:session_00042-1789017895";
constexpr std::uint64_t kMtime = 1789017895123456789ULL;

std::string Encoded(const rec::Command& cmd) {
    std::vector<std::uint8_t> out;
    EXPECT_TRUE(rec::Serialize(cmd, &out));
    return std::string(out.begin(), out.end());
}

rec::Result Decoded(const std::string& raw) {
    rec::Result r;
    EXPECT_TRUE(rec::DecodeResult(reinterpret_cast<const std::uint8_t*>(raw.data()), raw.size(), &r));
    return r;
}

// ---- vectors -------------------------------------------------------------------

TEST(RecordingsVectors, EveryCommandEncodesAsTheCorpusSays) {
    const auto v = visio_golden::Load("recordings_wire_vectors.txt");
    std::unique_ptr<rec::Command> cmd(new rec::Command);

    ASSERT_TRUE(rec::ListRecordingsCommand(cmd.get()));
    EXPECT_EQ(visio_golden::Hex(Encoded(*cmd)), visio_golden::Hex(v.at("list.cmd")));

    ASSERT_TRUE(rec::ListRecordingsCommand(cmd.get(), 20, kCursor));
    EXPECT_EQ(visio_golden::Hex(Encoded(*cmd)), visio_golden::Hex(v.at("list_page.cmd")));

    ASSERT_TRUE(rec::ListRecordingsCommand(cmd.get(), 0, "", kSession, "GILABS-A"));
    EXPECT_EQ(visio_golden::Hex(Encoded(*cmd)), visio_golden::Hex(v.at("list_session.cmd")));

    ASSERT_TRUE(rec::OpenRecordingFileCommand(cmd.get(), kSession, "ego_0003.mcap"));
    EXPECT_EQ(visio_golden::Hex(Encoded(*cmd)), visio_golden::Hex(v.at("open.cmd")));

    ASSERT_TRUE(rec::OpenRecordingFileCommand(cmd.get(), kSession, "ego_0003.mcap", 100663296,
                                              281018368, kMtime, "GILABS-A"));
    EXPECT_EQ(visio_golden::Hex(Encoded(*cmd)), visio_golden::Hex(v.at("open_resume.cmd")));

    ASSERT_TRUE(rec::DeleteRecordingCommand(cmd.get(), kSession, "GILABS-A"));
    EXPECT_EQ(visio_golden::Hex(Encoded(*cmd)), visio_golden::Hex(v.at("delete.cmd")));
}

TEST(RecordingsVectors, EveryResultDecodes) {
    const auto v = visio_golden::Load("recordings_wire_vectors.txt");

    const rec::Result open = Decoded(v.at("result_open.result"));
    EXPECT_EQ(open.command_id, 7u);
    EXPECT_TRUE(open.ok);
    ASSERT_EQ(open.payload, rec::Result::Payload::FileOpen);
    EXPECT_EQ(open.file_open.port, 50002u);
    EXPECT_EQ(open.file_open.offset, 100663296u);
    EXPECT_EQ(open.file_open.length, 180355072u);
    EXPECT_EQ(open.file_open.file_size, 281018368u);
    EXPECT_EQ(open.file_open.mtime_ns, kMtime);

    const rec::Result files = Decoded(v.at("result_session_files.result"));
    ASSERT_EQ(files.payload, rec::Result::Payload::Recordings);
    ASSERT_EQ(files.sessions.size(), 1u);
    const rec::SessionInfo& s = files.sessions[0];
    EXPECT_EQ(s.name, kSession);
    EXPECT_TRUE(s.active);
    EXPECT_EQ(s.file_count, 3u);
    EXPECT_DOUBLE_EQ(s.duration_s, 12.5);
    ASSERT_EQ(s.files.size(), 3u);
    EXPECT_EQ(s.files[0].name, "ego_0000.mcap");
    EXPECT_FALSE(s.files[0].writing);
    EXPECT_TRUE(s.files[0].complete);
    EXPECT_EQ(s.files[1].name, "ego_0001.mcap");
    EXPECT_TRUE(s.files[1].writing);
    EXPECT_FALSE(s.files[1].complete);
    EXPECT_EQ(s.files[2].name, "session.json");
    EXPECT_TRUE(s.files[2].writing);
    EXPECT_EQ(s.files[2].mtime_ns, kMtime + 2);

    const rec::Result page = Decoded(v.at("result_page.result"));
    ASSERT_EQ(page.payload, rec::Result::Payload::Recordings);
    ASSERT_EQ(page.sessions.size(), 2u);
    EXPECT_TRUE(page.sessions[1].damaged);
    EXPECT_EQ(page.next_cursor, kCursor);
    EXPECT_TRUE(page.read_while_recording);

    const rec::Result refused = Decoded(v.at("result_refused.result"));
    EXPECT_FALSE(refused.ok);
    EXPECT_EQ(refused.error_code, "writing");
    EXPECT_EQ(refused.error_message, "ego_0001.mcap is being recorded");
    EXPECT_EQ(refused.payload, rec::Result::Payload::None);
}

TEST(RecordingsBuilders, RefuseWhatTheDeviceCannotDecode) {
    std::unique_ptr<rec::Command> cmd(new rec::Command);
    EXPECT_FALSE(rec::OpenRecordingFileCommand(cmd.get(), "", "a.mcap"));
    EXPECT_FALSE(rec::OpenRecordingFileCommand(cmd.get(), ".hidden", "a.mcap"));
    EXPECT_FALSE(rec::OpenRecordingFileCommand(cmd.get(), "a/b", "a.mcap"));
    EXPECT_FALSE(rec::OpenRecordingFileCommand(cmd.get(), kSession, std::string(64, 'x')));
    EXPECT_TRUE(rec::OpenRecordingFileCommand(cmd.get(), std::string(63, 's'), std::string(63, 'f')));
    EXPECT_FALSE(rec::DeleteRecordingCommand(cmd.get(), kSession, ""));
    EXPECT_FALSE(rec::ListRecordingsCommand(cmd.get(), 0, std::string(64, 'c')));
}

// ---- pulling against a loopback sender ----------------------------------------------

class FakeDevice {
public:
    explicit FakeDevice(std::string data) : data_(std::move(data)) {
        listener_ = socket(AF_INET, SOCK_STREAM, 0);
        sockaddr_in addr{};
        addr.sin_family = AF_INET;
        addr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
        bind(listener_, reinterpret_cast<sockaddr*>(&addr), sizeof(addr));
        listen(listener_, 4);
        socklen_t len = sizeof(addr);
        getsockname(listener_, reinterpret_cast<sockaddr*>(&addr), &len);
        port_ = ntohs(addr.sin_port);
        thread_ = std::thread([this] { Serve(); });
    }
    ~FakeDevice() {
        stop_ = true;
        shutdown(listener_, SHUT_RDWR);
        close(listener_);
        thread_.join();
    }

    rec::Io Io(std::string* sink) {
        rec::Io io;
        io.run = [this](rec::Command* cmd, double, rec::Result* out) { return Run(*cmd, out); };
        io.sink = [sink](std::uint64_t offset, const std::uint8_t* d, std::size_t n) {
            EXPECT_EQ(offset, sink->size());
            sink->append(reinterpret_cast<const char*>(d), n);
            return true;
        };
        return io;
    }

    std::vector<std::uint64_t> cuts;  // bytes to send on the next connections
    bool refuse_sends = false;
    std::uint64_t mtime = kMtime;
    std::vector<std::uint64_t> open_offsets;
    std::vector<std::uint64_t> open_expect_mtimes;
    std::string data_;

private:
    bool Run(const rec::Command& cmd, rec::Result* out) {
        *out = rec::Result{};
        out->ok = true;
        if (cmd.which_body != visio_schema_v1_control_Command_open_recording_file_tag) {
            out->ok = false;
            out->error_code = "unsupported";
            return true;
        }
        const auto& b = cmd.body.open_recording_file;
        open_offsets.push_back(b.offset);
        open_expect_mtimes.push_back(b.expect_mtime_ns);
        if ((b.expect_size || b.expect_mtime_ns) &&
            (b.expect_size != data_.size() || b.expect_mtime_ns != mtime)) {
            out->ok = false;
            out->error_code = "changed";
            return true;
        }
        std::lock_guard<std::mutex> lock(mu_);
        std::uint64_t length = data_.size() - b.offset;
        pending_offset_ = b.offset;
        pending_length_ = length;
        pending_ = true;
        out->payload = rec::Result::Payload::FileOpen;
        out->file_open = rec::FileOpen{port_, b.offset, length, data_.size(), mtime};
        return true;
    }

    void Serve() {
        while (!stop_) {
            const int fd = accept(listener_, nullptr, nullptr);
            if (fd < 0) return;
            std::uint64_t off = 0, len = 0;
            bool have = false;
            {
                std::lock_guard<std::mutex> lock(mu_);
                have = pending_;
                pending_ = false;
                off = pending_offset_;
                len = pending_length_;
            }
            if (have && !refuse_sends) {
                if (!cuts.empty()) {
                    len = std::min(len, cuts.front());
                    cuts.erase(cuts.begin());
                }
                std::uint64_t sent = 0;
                while (sent < len) {
                    const ssize_t n = send(fd, data_.data() + off + sent, len - sent, MSG_NOSIGNAL);
                    if (n <= 0) break;
                    sent += static_cast<std::uint64_t>(n);
                }
                shutdown(fd, SHUT_WR);
            }
            close(fd);
        }
    }

    int listener_ = -1;
    std::uint16_t port_ = 0;
    std::thread thread_;
    std::atomic<bool> stop_{false};
    std::mutex mu_;
    bool pending_ = false;
    std::uint64_t pending_offset_ = 0;
    std::uint64_t pending_length_ = 0;
};

std::string Pattern(std::size_t n) {
    std::string s(n, '\0');
    for (std::size_t i = 0; i < n; ++i) s[i] = static_cast<char>((i * 31 + 7) & 0xFF);
    return s;
}

rec::PullRequest Request() {
    rec::PullRequest req;
    req.session_name = kSession;
    req.file_name = "ego_0000.mcap";
    return req;
}

TEST(RecordingsPull, WholeFileInOneOpen) {
    FakeDevice dev(Pattern(3 * 1024 * 1024 + 17));
    std::string got;
    const rec::Outcome o = rec::PullFile(dev.Io(&got), "127.0.0.1", Request());
    EXPECT_TRUE(o.ok) << o.code << " " << o.message;
    EXPECT_EQ(got, dev.data_);
    EXPECT_EQ(dev.open_offsets.size(), 1u);
    EXPECT_EQ(o.last_open.file_size, dev.data_.size());
}

TEST(RecordingsPull, EarlyCloseResumesAtTheBytesReceivedWithIdentityPinned) {
    FakeDevice dev(Pattern(2 * 1024 * 1024));
    dev.cuts = {100000, 300000};
    std::string got;
    const rec::Outcome o = rec::PullFile(dev.Io(&got), "127.0.0.1", Request());
    EXPECT_TRUE(o.ok) << o.code;
    EXPECT_EQ(got, dev.data_);
    EXPECT_EQ(dev.open_offsets, (std::vector<std::uint64_t>{0, 100000, 400000}));
    EXPECT_EQ(dev.open_expect_mtimes, (std::vector<std::uint64_t>{0, kMtime, kMtime}));
}

TEST(RecordingsPull, AFileThatChangesBetweenOpensFailsInsteadOfSplicing) {
    FakeDevice dev(Pattern(1024 * 1024));
    dev.cuts = {50000};
    std::string got;
    rec::Io io = dev.Io(&got);
    auto real_run = io.run;
    io.run = [&](rec::Command* cmd, double t, rec::Result* r) {
        if (!dev.open_offsets.empty()) dev.mtime += 1;
        return real_run(cmd, t, r);
    };
    const rec::Outcome o = rec::PullFile(io, "127.0.0.1", Request());
    EXPECT_FALSE(o.ok);
    EXPECT_EQ(o.code, "changed");
    EXPECT_EQ(got.size(), 50000u);
}

TEST(RecordingsPull, OpensThatDeliverNothingGiveUp) {
    FakeDevice dev(Pattern(4096));
    dev.refuse_sends = true;
    std::string got;
    rec::Options opt;
    opt.max_opens_without_progress = 3;
    const rec::Outcome o = rec::PullFile(dev.Io(&got), "127.0.0.1", Request(), opt);
    EXPECT_EQ(o.code, "no_progress");
    EXPECT_EQ(dev.open_offsets.size(), 3u);
}

TEST(RecordingsPull, AFailingSinkStopsThePull) {
    FakeDevice dev(Pattern(1024 * 1024));
    std::string got;
    rec::Io io = dev.Io(&got);
    io.sink = [](std::uint64_t, const std::uint8_t*, std::size_t) { return false; };
    const rec::Outcome o = rec::PullFile(io, "127.0.0.1", Request());
    EXPECT_EQ(o.code, "sink");
    EXPECT_EQ(dev.open_offsets.size(), 1u);
}

TEST(RecordingsPull, AnAlreadyCompleteFileNeedsNoSocket) {
    FakeDevice dev(Pattern(512));
    std::string got;
    rec::PullRequest req = Request();
    req.offset = 512;
    const rec::Outcome o = rec::PullFile(dev.Io(&got), "127.0.0.1", req);
    EXPECT_TRUE(o.ok);
    EXPECT_EQ(o.last_open.length, 0u);
}

TEST(RecordingsList, FollowsCursorsAndRefusesOneThatDoesNotAdvance) {
    int calls = 0;
    rec::Io io;
    io.run = [&](rec::Command* cmd, double, rec::Result* r) {
        *r = rec::Result{};
        r->ok = true;
        r->payload = rec::Result::Payload::Recordings;
        const std::string cursor = cmd->body.list_recordings.cursor;
        rec::SessionInfo s;
        s.name = "session_" + std::to_string(calls++);
        r->sessions.push_back(s);
        if (cursor.empty()) r->next_cursor = "p2";
        return true;
    };
    std::vector<rec::SessionInfo> out;
    std::string code;
    ASSERT_TRUE(rec::ListAllRecordings(io.run, "", &out, &code)) << code;
    EXPECT_EQ(out.size(), 2u);

    io.run = [&](rec::Command*, double, rec::Result* r) {
        *r = rec::Result{};
        r->ok = true;
        r->payload = rec::Result::Payload::Recordings;
        r->next_cursor = "same";
        return true;
    };
    EXPECT_FALSE(rec::ListAllRecordings(io.run, "", &out, &code));
    EXPECT_EQ(code, "protocol");
}

}  // namespace

TEST(RecordingsConstants, PullQuiesceRulesAreTheSpecList) {
  // Python PULL_QUIESCE_RULES and TS PULL_QUIESCE_RULES pin the same list.
  ASSERT_EQ(std::size(rec::kPullQuiesceRules), 4u);
  EXPECT_STREQ(rec::kPullQuiesceRules[0], "**/camera/*");
  EXPECT_STREQ(rec::kPullQuiesceRules[1], "**/imu/*/raw");
  EXPECT_STREQ(rec::kPullQuiesceRules[2], "**/imu/*/quat");
  EXPECT_STREQ(rec::kPullQuiesceRules[3], "**/audio/*");
}
